framework_tvb/tvb/core/services/project_service.py

```diff
@@ -521,65 +523,123 @@ def get_datatype_details(datatype_gid):
 
-            op = dao.get_operation_by_id(datatype.fk_from_operation)
             if links:
-                was_link = False
-                for link in links:
-                    # This means it's only a link and we need to remove it
-                    if link.fk_from_datatype == datatype.id and link.fk_to_project == project.id:
-                        dao.remove_entity(Links, link.id)
-                        was_link = True
-                if not was_link:
-                    # Create a clone of the operation
-                    # There is no view_model so the view_model_gid is None
-
-                    new_op = Operation(op.view_model_gid,
-                                       dao.get_system_user().id,
-                                       links[0].fk_to_project,
-                                       datatype.parent_operation.fk_from_algo,
-                                       datatype.parent_operation.status,
-                                       datatype.parent_operation.start_date,
-                                       datatype.parent_operation.completion_date,
-                                       datatype.parent_operation.fk_operation_group,
-                                       datatype.parent_operation.additional_info,
-                                       datatype.parent_operation.user_group,
-                                       datatype.parent_operation.range_values)
-                    new_op = dao.store_entity(new_op)
-                    to_project = self.find_project(links[0].fk_to_project)
-                    to_project_path = self.storage_interface.get_project_folder(to_project.name)
-
-                    full_path = h5.path_for_stored_index(datatype)
-                    old_folder = self.storage_interface.get_project_folder(project.name, str(op.id))
-                    vm_full_path = h5.determine_filepath(op.view_model_gid, old_folder)
-
-                    self.storage_interface.move_datatype_with_sync(to_project, to_project_path, new_op.id, full_path,
-                                                                   vm_full_path)
-
-                    datatype.fk_from_operation = new_op.id
-                    datatype.parent_operation = new_op
-                    dao.store_entity(datatype)
-                    dao.remove_entity(Links, links[0].id)
+                op = dao.get_operation_by_id(datatype.fk_from_operation)
+                # Instead of deleting, we copy the datatype to the linked project
+                # We also clone the operation
+                new_operation = self.__copy_linked_datatype_before_delete(op, datatype, project,
+                                                                          links[0].fk_to_project)
+
+                # If there is a  datatype group and operation group and they were not moved yet to the linked project,
+                # then do it
+                if datatype.fk_datatype_group is not None:
```

Comment:
isn't this called for every dt in a group? Could we avoid that?

Reply:
No it is not. The condition you asked before at line 673 makes the flow not reach this method for every datatype in the group.